### PR TITLE
Add `Field::new_list_field` and improve `DataType::new_list` docs

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Re https://github.com/apache/arrow-rs/issues/4544

# Rationale for this change
 
As described on https://github.com/apache/arrow-rs/pull/4561#issuecomment-1661940270 the API for creating List fields is somewhat confusing and hiding details of doing so may make it more so. 

# What changes are included in this PR?
1.  Add ~`Field::new_list_item`~ `Field::new_list_field`
2. improve `DataType::new_list` docs

# Are there any user-facing changes?
Yes a new API for constructing `Field`s and better docs

### CODE REVIEW

File: arrow-schema/src/field.rs [hunk 1]
```diff
@@ -129,6 +129,24 @@ impl Field {
...
+    ///
+    /// While not required, by convention the inner `Field` of these
+    /// types is named `"item"`
+    ///
+    /// # Example
+    /// ```
+    /// # use arrow_schema::{Field, DataType};
+    /// assert_eq!(
+    ///   Field::new("item", DataType::Int32, true),
+    ///   Field::new_list_item(DataType::Int32, true)
```
Comment (tustvold) [inline arrow-schema/src/field.rs:143]: It does occur to me that this is now longer :sweat_smile:

**Review by tustvold — APPROVED** [review]
Not sure about new_list_item, docs on DataType::new_list look good

File: arrow-schema/src/field.rs (hunk 1 above)
Comment (alamb) [reply arrow-schema/src/field.rs:143]: But it is more explict / discoverable

Maybe `Field::new_list_field`? or `Field::new_list`?

**Comment by alamb** [comment]
I'll give this PR a few more days to see if anyone else has thoughts / comments

File: arrow-schema/src/datatype.rs [hunk 2]
```diff
@@ -577,9 +577,14 @@ impl DataType {
         }
     }
 
-    /// Create a List DataType default name is "item"
+    /// Create a [`DataType::List`] where each elements has the
```
Comment (viirya) [inline arrow-schema/src/datatype.rs:580]: Not sure 😄 but maybe?

```suggestion
    /// Create a [`DataType::List`] where each element has the
```

File: arrow-schema/src/datatype.rs [hunk 3]
```diff
@@ -577,9 +577,14 @@ impl DataType {
...
+    /// specified type and nullability and coventional name
```
Comment (viirya) [inline arrow-schema/src/datatype.rs:581]:
```suggestion
    /// specified type and nullability and conventional name
```

File: arrow-schema/src/field.rs (hunk 1 above)
Comment (viirya) [reply arrow-schema/src/field.rs:143]: I guess `new_list_field`? As this is to create a field.

**Review by viirya — APPROVED** [review]

File: arrow-schema/src/field.rs (hunk 1 above)
Comment (alamb) [reply arrow-schema/src/field.rs]: Great idea -- done in 76b1e70896

**Comment by tustvold** [comment]
How should we proceed with this one?

**Comment by alamb** [comment]
I will polish it up and get it merged

**Pull request merged**